Make notebooks appear well in the dark theme

patalvarado opened the pull request:
Modify the IPYNB files to remove hard-coded text colors.

adamtwo commented:
It looks like the sed command doesn't resolve all the issues, e.g. `/lab/tree/UseCases/Anomaly_Detection/Anomaly_Detection_Python_TrustedAI.ipynb` looks like this.

I'm not sure if there is any other way to validate, but to each notebook and scroll through it to make sure it looks good.

<img width="1150" alt="image" src="https://github.com/user-attachments/assets/c30af7d2-820f-430c-ac3f-84722df9fec4" />

DallasBowden commented:
The search and replace is also leaving the tops of some of the notebooks looking like this:

![image](https://github.com/user-attachments/assets/abb8adaa-6c18-41a5-9d2a-55690cedb9fb)

patalvarado commented:
I'll examine all files next week to reconcile the issues and validate each one. Thanks!

DallasBowden commented:
Here are my initial recommendations.  If you're a linux expert you can probably do it with a few passes using sed.  Or, It shouldn't be very difficult to write some python to do this:

1. Leave the Header with the graphic alone.  It shows well either way.
2. Update the footer by removing 'background:#f9f9f9' and '#00233C' from the first line.
3. Update the line tag in markdown cells to only be <hr>.
4. Remove 'background-color:#00233C' and 'color:02233c' from all other markup cells.
 
This will be a good start.  I haven't looked at any of the notebooks with a lot of graphics to see how they display when the view mode is changed.

patalvarado commented:
Great! I'll start with that and revert back the originals and use grep to find all the lines that have the color hardcoded, then compose the sed commands to fix.

Thanks!

DougEbel commented:
we should be using CSS.  Marketing makes changes to "standard marketing
fonts and colors" on a routine basis and these definitions should be in one
place.  Plus, hard coding into each of the hundreds of jupyter notebooks is
not a good approach and may not be consistent.

On Sat, Jan 18, 2025 at 11:35 AM Pat Alvarado ***@***.***>
wrote:

> Great! I'll start with that and revert back the originals and use grep to
> find all the lines that have the color hardcoded, then compose the sed
> commands to fix.
>
> Thanks!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Teradata/jupyter-demos/pull/809#issuecomment-2599779084>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AYNBXG2WF6YB74M4TZPUA232LJ7FJAVCNFSM6AAAAABVM5Y2JCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKOJZG43TSMBYGQ>
> .
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>

patalvarado commented:
I agree we should use Cascading Style Sheets (CSS). For this initial fix I'd like to get this done with these direct edits first for expediency and then I will look into establishing and documenting a default style format in a single css file that can be applied to every notebook. Additionally documenting guidelines for notebook authors to ensure the notebooks support both light and dark themes since an author can override a parent's style class with another class. I'm working on a notebook using one of the Use Case notebooks as a template so I can experiment on setting up and testing the css file and then determine how to apply the changes to all the existing notebooks.

DougEbel commented:
Dallas Bowden and four people in the GDC are creating/updating 179 demos. Most of the notebook files have read-only versions, yielding 299 entries in the demo_files table. It is a good idea to coordinate notebook changes with the content team since both you and the content team could have changes to the same demo(s) on different branches.

patalvarado commented:
I complete the updates, the majority taken care of with sed. I spent most of the day reviewing each notebook to make sure they are viewable in both themes and made some manual edits of a few.

## Stream Edits using sed
### Remove hardcoded text color on any element with the color property set to #00233C or #00233c to revert to the default color
`find . -type f -name "*.ipynb" -exec sed -i -E 's/(style *= *[^>]*)(; *color *: *#00233c)/\1/gI' {} +`
### Remove hardcoded text color of #00233C or #00233c from the horizontal rule to revert to the default color
`find . -type f -name "*.ipynb" -exec sed -i -E 's/(<hr *style *= *[^>]*)(; *background-color *: *#00233C)/\1/gI' {} +`
### Replace the bottom border line color of the footer with a color that will appear in both dark and light theme
`find . -type f -name "*.ipynb" -exec sed -i -E 's/(<footer *style.*)(; *background *: *#f9f9f9)( *; *border-bottom *: *.*px *solid *)( *#00233C)/\1\3#91A0Ab/gI' {} +`
### Remove the light background in the span element and apply a bold as well as italic to the font
`find . -type f -name "*.ipynb" -exec sed -i -E 's/(<span *[^>]*)(background-color *: *#eee; *)/\1font-weight: bold; /gI' {} +`

## Manual Edits
### Getting_Started/PySpark_to_Teradataml/HousingPrices_pyspark_to_tdml.ipynb
1. In section 4. Understanding the HTML report there are \<code\> elements that set the text color to match the color that would appear in the HTML report which follows the theme. For red and blue it looks fine, but for black it gets washed out in Dark Theme because the color turns to white. I added a light gray background color and then edited the text to explain that the normal text may appear white in the Dark Theme.

### Getting_Started/Teradataml_Widgets/Plot_Notebook.ipynb
1. In the section titled "Plot Notebook Toolbar Button Functionality" there is a <div> element with the background color set to background-color:#f5f5f5 which is nearly white, so it washes out the white text in Dark Mode, so I changed it to background-color:gray
2. In the section titled "Class: plot.Ui" there is a <div> element with the background color set to background-color:#f5f5f5 which is nearly white, so it washes out the white text in Dark Mode, so I changed it to background-color:gray

### Getting_Started/Teradataml_Widgets/Tutorial.ipynb
1. In the section titled "Notebook Toolbar Button Functionality" there is a <div> element with the background color set to background-color:#f5f5f5 which is nearly white, so it washes out the white text in Dark Mode, so I changed it to background-color:gray
2. In the section titled "Class: analytioc_functions.Ui" there is a <div> element with the background color set to background-color:#f5f5f5 which is nearly white, so it washes out the white text in Dark Mode, so I changed it to background-color:gray

### ModelOps/11_ModelOps_GIT_PIMA_Python_indb_XGboost.ipynb
1. The footer used a light background and copyright text different from the other washing out the text in light mode. I replaced it with the same format as the footer of the other ModelOps notebooks.

### ModelOps/13_ModelOps_GIT_PIMA_Python_osml_DecisionTree.ipynb
1. The footer used a light background and copyright text different from the other washing out the text in light mode. I replaced it with the same format as the footer of the other ModelOps notebooks.

### ModelOps/14_ModelOps_GIT_PIMA_Python_osml_LightGBM.ipynb
1. The footer used a light background and copyright text different from the other washing out the text in light mode. I replaced it with the same format as the footer of the other ModelOps notebooks.

### Recipes/Feature_Store_Feature_Engg/Feature_Engineering_and_Feature_Store_Usage.ipynb
1. The image for "The feature store approach" had a transparent background. I add a style = 'background-color:white' to provide a white background to accomodate both Light and Dark Theme.

### UseCases/AnomalyDetection_OutstandingAmount_FS/AnomalyDetection_OutstandingAmount_FeatureStore.ipynb
1. The image for "The feature store approach" had a transparent background. I add a style = 'background-color:white' to provide a white background to accomodate both Light and Dark Theme.

### UseCases/Complaints_Analysis_Azure_OpenAI/Complaint_Analysis_Dashboard.ipynb
1. The background-color on each of the sections containing an image was set to #F5F5F5. I removed it to adapt to both Dark and Light modes.
2. The border around the <div> block was set to the dark color #00233C. I changed it to color #91A0Ab that would work for both Dark and Light themes.

### UseCases/Complaints_Analysis_GenAI_Bedrock/Complaint_Analysis_Dashboard.ipynb
1. The background-color on each of the sections containing an image was set to #F5F5F5. I removed it to adapt to both Dark and Light modes.
2. The border around the <div> block was set to the dark color #00233C. I changed it to color #91A0Ab that would work for both Dark and Light themes.

### UseCases/Complaints_Analysis_GenAI_Gemini/Complaint_Analysis_Dashboard.ipynb
1. The background-color on each of the sections containing an image was set to #F5F5F5. I removed it to adapt to both Dark and Light modes.
2. The border around the <div> block was set to the dark color #00233C. I changed it to color #91A0Ab that would work for both Dark and Light themes.

### UseCases/EFS_Demo/Telco_Churn_using_EFS.ipynb
1. The image for "The feature store approach" had a transparent background. I add a style = 'background-color:white' to provide a white background to accomodate both Light and Dark Theme.

### UseCases/EFS_Demo/Telco_Churn_using_tdfs4ds_Feature_Store.ipynb
1. The image for "The feature store approach" had a transparent background. I add a style = 'background-color:white' to provide a white background to accomodate both Light and Dark Theme.

### UseCases/Energy_Consumption_Forecasting_AzureML/Energy_Consumption_Forecasting_AzureML.ipynb
1. The background for the warning below the section "1.1 Downloading and installing additional software needed" was changed from #FFdddd to #FF7777 to adapt to both Dark and Light Theme.
2. The background for the warning below the section "7.1 Work Tables" was changed from #FFdddd to #FF7777 to adapt to both Dark and Light Theme.

### UseCases/Store_Sales_Forecasting_Prophet_STO/Store_Sales_Forecasting_Prophet_STO_PY_SQL.ipynb
1. The image in the first "Introduction" section depicting the AMPs and Python program had a transparent background. I add a style = 'background-color:white' to provide a white background to accomodate both Light and Dark Theme.

### UseCases/Supply_Chain_Visibility/Supply_Chain_Visibility_Python.ipynb
1. The footer used a light background and copyright text different from the other washing out the text in light mode. I replaced it with the same format as the footer of the other ModelOps notebooks.

## Recommendations
While I was checking all the notebooks for viewing in both Light and Dark themes I came across some that caught my attention for potential recommendations.

### Query Service Notebook
The Query Service Notebook provides a manual way of switching to the dashboard to copy the hostname, but the hostname is stored within the VM, so I recommend adding this Python code to retrieve and display for copying or linking:
```
hostname_path = '/home/jovyan/JupyterLabRoot/.usersettings/hostname'
with open(hostname_path, 'r') as file:
    csae_hostname = file.read().strip()

query_service_url_http = 'http://' + csae_hostname + ':1080'
query_service_url_https = 'https://' + csae_hostname + ':1443'

print("Query Service HTTP URL: {}/\n".format(query_service_url_http))
print("Query Service HTTPS URL: {}/".format(query_service_url_https))
```
### Index
1. Recommendation to add estimated time to run through each notebook to set expectations of the time investment.
2. Include a filter that lists all the ClearScape and Vantage Analytics Library functions/algorithms so that a developer looking for examples on how to use a certain function can find all the notebooks that have them as examples.

patalvarado commented:
> Dallas Bowden and four people in the GDC are creating/updating 179 demos. Most of the notebook files have read-only versions, yielding 299 entries in the demo_files table. It is a good idea to coordinate notebook changes with the content team since both you and the content team could have changes to the same demo(s) on different branches.

Hi Doug, my changes are primarily edits to the Markdown cells so conflicts may arise if changes are made to the markdown code or the text within the markdown code being changed. Other than Dallas and Adam, who on the content team should I coordinate? Dallas provided the recommendations which is what I applied in my changes for the most part. I documented all the other manual changes in my comments after I committed my changes which involved some images with transparent backgrounds and some background color adjustments.

adamtwo commented:
> > Dallas Bowden and four people in the GDC are creating/updating 179 demos. Most of the notebook files have read-only versions, yielding 299 entries in the demo_files table. It is a good idea to coordinate notebook changes with the content team since both you and the content team could have changes to the same demo(s) on different branches.
> 
> Hi Doug, my changes are primarily edits to the Markdown cells so conflicts may arise if changes are made to the markdown code or the text within the markdown code being changed. Other than Dallas and Adam, who on the content team should I coordinate? Dallas provided the recommendations which is what I applied in my changes for the most part. I documented all the other manual changes in my comments after I committed my changes which involved some images with transparent backgrounds and some background color adjustments.

IMO, our standard git workflow doesn't require additional coordination as even when there is a conflict, it should be all trivial git merges that git should do for us. @DallasBowden , if you could please let the team know that a large PR might be coming and that they will need to either rebase or merge from main to their branch, I think that will do it.

@DallasBowden I'm not familiar with the process of regenerating the static notebooks. Now that we have playwright, perhaps it's something that we want to automate, regenerate with each merge to main?

@patalvarado, I'm about to review the PR. I'll leave the comments in the PR tool.

adamtwo commented:
@patalvarado , the PR contains all the files in the repo. As far as I can tell, that's because the PR attempts to modify file permissions. Let's fix that first as the PR tool crashes in my browser with 994 files in scope :).
The issue is likely due to you running your git client on Windows. I believe this can help you solve the issue: https://stackoverflow.com/questions/6476513/git-file-permissions-on-windows. 

Another challenge that we will have is how to make sure that the PR doesn't break anything. We have nightly tests that we can run on your PR to validate the functionality. 

We also need to validate that the new appearance is ok. It's tru for both the light mode (to make sure we don't break it) and the dark mode (to make sure that it's of good quality).

How can we ensure that the appearance is good? Can we do something quicker than manually inspecting all the notebooks in 2 themes?

patalvarado commented:
I removed the executable flag from all files except those with the .sh extension which has the executable flag set on the original files in my CSAE environment. All git commands executed from Linux, not using Windows for pushing or committing changes.

patalvarado commented:
> @patalvarado , the PR contains all the files in the repo. As far as I can tell, that's because the PR attempts to modify file permissions. Let's fix that first as the PR tool crashes in my browser with 994 files in scope :). The issue is likely due to you running your git client on Windows. I believe this can help you solve the issue: https://stackoverflow.com/questions/6476513/git-file-permissions-on-windows.
> 
> Another challenge that we will have is how to make sure that the PR doesn't break anything. We have nightly tests that we can run on your PR to validate the functionality.
> 
> We also need to validate that the new appearance is ok. It's tru for both the light mode (to make sure we don't break it) and the dark mode (to make sure that it's of good quality).
> 
> How can we ensure that the appearance is good? Can we do something quicker than manually inspecting all the notebooks in 2 themes?

Hi Adam, I made the updates to the modes using the instructions from those links though GitHub still records 944 files changed.

I can try resyncing my repository main branch to the latest and then creating a new branch from the main and repeat the process solely using the GitHub web site and Linux git and gh commands and then compare permission modes before and after on all the files are committed.

The only way to validate the dark theme and light theme is through manual review.

I propose just updating a few notebooks at a time and collaborate a review with the team on Microsoft Teams or a subset of the team in 15-to-30-minute review sessions. As a preview, we can get on a short call and I can show you the changes made on my branch as they appear on Clearscape Analytics Experience.

For example, 

Phase 1:
- FeaturePreview: 1 notebook
- Getting_Started: 20 notebooks
- ModelOps: 17 notebooks
- Total: 38 notebooks to review

Phase 2:
- Recipes: 15 notebooks
- Resume_Analyzer: 3 notebooks
- Teradata: 2 notebooks
- Total: 20 notebooks to review

Phase 3:
- UseCases: 4D_Analytics_on_bike_sharing to Chat_with_docs_GenAI: 26 notebooks

Phase 4:
- UseCases: Competitor_Proximity_Geospatial_Analysis to Complaints_Analysis_GenAI_Gemini: 26 notebooks

Phase 5:
- UseCases: Consumer_Complaints to Energy_Consumption_Forecasting: 26 notebooks

Phase 6:
- UseCases: Financial_Customer_Journey to Language_Models_InVantage: 28 notebooks

Phase 7:
- UseCases: Marketing_Campaign_Effectiveness to Signal_Processing_and_Classification: 27 notebooks

Phase 8:
- UseCases: Store_Sales_Forecasting_ARIMA to Vertex_AI_Integration + startup.ipynb: 20 notebooks

Total Review Time: 15 to 30 minutes each phase for a max of ~ 4 hours

Let me know what you think,

Thanks!

DallasBowden commented:
Replace this PR with PR #830.

DallasBowden commented:
Replace this PR with PR #830.
